#169257402 Profile page settings

gildniy opened the pull request:
#### What does this PR do?
Create Profile page settings for the user
#### Description of Task to be completed?
- [x] Create user profile information at signup (default: `firstName`, `lastName`, `email`, `role`).
- [x] Others fields in the User model which are set null at this stage are: `birthDate`, `preferedLanguage`, `residentAddress`, `department`, `preferredCurrency`, `gender`, `lastLogin`, `phoneNumber` and `lineManagerId`

#### How should this be manually tested?
- Clone the application to the local machine
- Install the dependencies  by `npm install`
- Run `npm run migrate` to create the database tables
- Run `npm run watch` to run the server
- Open postman, enter in the address bar `http://localhost:3000/api/v1/auth/signup` with method POST and create a new user by entering an object having the `firstName`, `lastName`, `email` and `password` like and hit on `SEND` button:
```
{
	"firstName": "Gildas",
	"lastName": "Niyigena",
	"email": "gildniy05@gmail.com",
	"password": "passwor6"
}
```
- The user will be created with a profile.
- By going to the route `PATCH /api/v1/user/update-profile`,  and filling in this data:
```
{
	"firstName": "John",
    "lastName": "Dor",
    "email": "john99@barefoot.com",
    "password": "p4$$W0rd",
    "birthDate": "1950-01-01T00:00:00.000Z",
    "preferredLanguage": "english",
    "preferredCurrency": "$",
    "residenceAddress": "Kibungo",
    "gender": "male",
    "department": "markerting",
    "lineManagerId": 2,
    "phoneNumber": "000-000-0012"
}
```
You will see the user information in the body, as an updated profile information

- By going to the route `GET /api/v1/user/1`, you should see the user profile information

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#169257402](https://www.pivotaltracker.com/story/show/169257402)
#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

gildniy commented on src/models/profile.js:
```diff
@@ -0,0 +1,32 @@
+    lineManagerId: DataTypes.INTEGER,
+    userId: DataTypes.INTEGER
+  }, {});
+  Profile.associate = function (models) {
```
Fixed

nshaphan commented:
Removing Names from the user model would require many changes to the merged and unmerged PRs, I think that you can create a separate model for other user profile information

gildniy commented:
> Removing Names from the user model would require many changes to the merged and unmerged PRs, I think that you can create a separate model for other user profile information

You're right, I'll display the name from the tables relationship

gildniy committed:
feature(profile) user profile setting
 - upon registration user profile will be created
 - user can view profiles
 - user can edit their own profile
[Finishes #169861079]

gildniy committed:
feature(profile) user profile setting
 - upon registration user profile will be created
 - user can view profiles
 - user can edit their own profile
[Finishes #169861079]

nshaphan commented on src/migrations/20191112101055-create-user.js:
```diff
@@ -7,28 +7,81 @@ module.exports = {
+    lineManager: {
+      allowNull: true,
+      type: Sequelize.STRING,
+    },
```
Great work @gildniy, but I think this would have a type of integer and be a foreign key linking users and their manager

gildniy replied:
I made this thinking a case where the line manager could be a non-registered user.

nshaphan replied:
I think that all the manager will be registered users on the platform

gildniy replied:
Ok you're right, let me implement it that way, thanks for the suggestion.

RaymondMwaura commented on src/routes/api/users.js:
```diff
@@ -247,8 +252,99 @@ router.post('/auth/forgotPassword', catchErrors(users.forgotPassword));
  *        description: email not found
  *      '200':
- *        description: succesfull sent reset link
+ *        description: successful sent reset link
```
I think you should change this to: 'successfully sent reset link'

RaymondMwaura commented on src/routes/api/users.js:
```diff
@@ -206,7 +211,7 @@ router.get('/auth/reverifyUser', catchErrors(users.resendEmail));
  *        description: email not found
  *      '200':
- *        description: succesfull sent reset link
+ *        description: successful sent reset link
```
I think you should change this to: 'successfully sent reset link'

RaymondMwaura commented on src/routes/api/users.js:
```diff
@@ -165,7 +170,7 @@ router.post('/auth/signin', validation, catchErrors(users.findUser));
  *        description: email not found
  *      '200':
- *        description: succesfull verified
+ *        description: successful verified
```
I think you should change this to: 'successfully verified'

RaymondMwaura commented on src/routes/api/users.js:
```diff
@@ -87,7 +92,7 @@ router.post('/auth/signup', validation, checkForEmail, catchErrors(users.createU
  *        description: trying to verify again
  *      '200':
- *        description: succesfull verified
+ *        description: successful verified
```
I think you should change this to: 'successfully verified'

RaymondMwaura commented on src/models/User.js:
```diff
@@ -1,10 +1,74 @@
-    isVerified: DataTypes.BOOLEAN,
-    password: DataTypes.STRING
+    firstName: {
+      allowNull: true,
```
I don't think we should allow this field to be nullable since signing up fills it as a requirement.

RaymondMwaura commented on src/models/User.js:
```diff
@@ -1,10 +1,74 @@
+      type: DataTypes.STRING
+    },
+    lastName: {
+      allowNull: true,
```
I don't think we should allow this field to be nullable since signing up fills it as a requirement.

nshaphan replied:
This works fine, great job @gildniy, update your description. update Pull request description and work on @RaymondMwaura  feedback

RaymondMwaura commented on src/models/User.js:
```diff
@@ -15,6 +79,14 @@ export default (sequelize, DataTypes) => {
+      foreignKey: 'lineManagerId',
+      onDelete: 'CASCADE'
+    });
+    User.belongsTo(models.user, {
```
I think this association is unnecessary since we don't have a separate 'manager' model.

RaymondMwaura replied:
I think this association is unnecessary since we don't have a separate 'manager' model.

RaymondMwaura reviewed (changes requested):
I think this is overall a great job. I have some feedback for you to consider.

I think you should change the error message when a non-existent id is given (e.g. ‘Profile doesn’t exist’). Currently, it returns the following:

http://localhost:3000/api/v1/user/<non_existent_id>/get-profile

```
{
    "status": "error",
    "message": "Cannot read property 'dataValues' of null"
}
```
I find the required phone number format to be unclear -> the error message doesn’t clearly show a user the required response format. I found it a challenge to enter a Rwandan number: 256 786 123 456. Maybe you can change the regex rule here to be more flexible.
```
{
    "status": "error",
    "message": [
        "phoneNumber with value \"256-786-123456\" fails to match the required pattern: /^(\\(\\d{3}\\) |\\d{3}-)\\d{3}-\\d{4}$/"
    ]
}
```
I think you should remove the information about creating a separate profile model from the PR description since you switched to using only the User model by updating it with the required fields.

How will a user’s line manager be updated?
```
{
    "status": "error",
    "message": [
        "\"lineManager\" is not allowed"
    ]
}
```
I think you might have to create a separate route specifically for updating a user’s line manager since the current update profile route (PATCH /api/v1/user/update-profile) doesn’t allow you to update a user’s line manager and then maybe not allow a user to update his/her own manager.

nshaphan reviewed (dismissed):
Great work @gildniy

gildniy commented:
> I think this is overall a great job. I have some feedback for you to consider.
> 
> I think you should change the error message when a non-existent id is given (e.g. ‘Profile doesn’t exist’). Currently, it returns the following:
> 
> http://localhost:3000/api/v1/user/<non_existent_id>/get-profile
> 
> ```
> {
>     "status": "error",
>     "message": "Cannot read property 'dataValues' of null"
> }
> ```
> 
> I find the required phone number format to be unclear -> the error message doesn’t clearly show a user the required response format. I found it a challenge to enter a Rwandan number: 256 786 123 456. Maybe you can change the regex rule here to be more flexible.
> 
> ```
> {
>     "status": "error",
>     "message": [
>         "phoneNumber with value \"256-786-123456\" fails to match the required pattern: /^(\\(\\d{3}\\) |\\d{3}-)\\d{3}-\\d{4}$/"
>     ]
> }
> ```
> 
> I think you should remove the information about creating a separate profile model from the PR description since you switched to using only the User model by updating it with the required fields.
> 
> How will a user’s line manager be updated?
> 
> ```
> {
>     "status": "error",
>     "message": [
>         "\"lineManager\" is not allowed"
>     ]
> }
> ```
> 
> I think you might have to create a separate route specifically for updating a user’s line manager since the current update profile route (PATCH /api/v1/user/update-profile) doesn’t allow you to update a user’s line manager and then maybe not allow a user to update his/her own manager.

Thanks @RaymondMwaura , let me work on these.

gildniy commented:
> > I think this is overall a great job. I have some feedback for you to consider.
> > I think you should change the error message when a non-existent id is given (e.g. ‘Profile doesn’t exist’). Currently, it returns the following:
> > http://localhost:3000/api/v1/user/<non_existent_id>/get-profile
> > ```
> > {
> >     "status": "error",
> >     "message": "Cannot read property 'dataValues' of null"
> > }
> > ```
> > 
> > 
> > I find the required phone number format to be unclear -> the error message doesn’t clearly show a user the required response format. I found it a challenge to enter a Rwandan number: 256 786 123 456. Maybe you can change the regex rule here to be more flexible.
> > ```
> > {
> >     "status": "error",
> >     "message": [
> >         "phoneNumber with value \"256-786-123456\" fails to match the required pattern: /^(\\(\\d{3}\\) |\\d{3}-)\\d{3}-\\d{4}$/"
> >     ]
> > }
> > ```
> > 
> > 
> > I think you should remove the information about creating a separate profile model from the PR description since you switched to using only the User model by updating it with the required fields.
> > How will a user’s line manager be updated?
> > ```
> > {
> >     "status": "error",
> >     "message": [
> >         "\"lineManager\" is not allowed"
> >     ]
> > }
> > ```
> > 
> > 
> > I think you might have to create a separate route specifically for updating a user’s line manager since the current update profile route (PATCH /api/v1/user/update-profile) doesn’t allow you to update a user’s line manager and then maybe not allow a user to update his/her own manager.
> 
> Thanks @RaymondMwaura , let me work on these.

Fixed now.

nshaphan reviewed (dismissed):
Great works @gildniy, this works as expected

RaymondMwaura reviewed (dismissed):
This is good work. All the tests are passing after the changes you made.

timolinn commented on src/controllers/users.js:
```diff
@@ -188,6 +190,61 @@ class UserController {
+
+      if (data === 'own_manage') {
+        return Responses.handleError(409, 'You cannot manage your self', res);
+      } else {
```
No need for this else statement. You can simple return the success handler outside the `if` block.

gildniy replied:
You're right, let me change it that way.

timolinn commented on src/controllers/users.js:
```diff
@@ -188,6 +190,61 @@ class UserController {
+        const { password, lineManagerId, ...data } = user;
+
+        return Responses.handleSuccess(200, 'Profile retrieved!', res, data);
+      } else {
```
no need for this `else` statement.

gildniy replied:
You're right, let me change it that way.

timolinn commented on src/migrations/20191112101055-create-user.js:
```diff
@@ -7,28 +7,81 @@ module.exports = {
-      type: Sequelize.STRING
+      allowNull: false,
+      type: Sequelize.STRING,
+      defaultValue: 'password',
```
Why do we have a default password value? I think this is insecure.

gildniy replied:
This was made to let the user registering using Facebook or Google to be accepted, as the password field was required. If this remove, the Social Registration of the user will fail.

timolinn replied:
If a user signs up with google, and tries to log in with his gmail and the default `password` as password, would the user be authenticated?

timolinn commented on src/routes/api/users.js:
```diff
@@ -247,8 +252,99 @@ router.post('/auth/forgotPassword', catchErrors(users.forgotPassword));
+ *      '500':
+ *        description: Internal Server error
+ * */
+router.get('/user/:userId/get-profile', verifyUser, catchErrors(getUserProfile));
```
I don't think `/get-profile` is necessary here, `/user/:userId` should do.

gildniy replied:
Ok thanks, let me change the route that way.

timolinn commented on src/services/User.service.js:
```diff
@@ -9,13 +9,121 @@ class UserServices {
+      });
+    }
+
+    return { userData, lineManagerData };
```
I believe there's a way to make sequelize fetch the lineManager, provided the lineManagerId was set as foreign key, we can add the code below to `findOne` query to preload all the data including the relationships:
```javascript
include:[{
  all: true
}]
```

gildniy replied:
Fixed, I used 
```
include: ['LineManager']
```
instead.

timolinn commented on src/services/User.service.js:
```diff
@@ -9,13 +9,121 @@ class UserServices {
+   * @param { object } res response object
+   * @returns {object} return the user's data
+   */
+  async updateUserInfoByEmail(attribute, email) {
```
change function parameter to `attributes`.

gildniy replied:
Let me change it that way, thanks.

gildniy commented:
> > I think this is overall a great job. I have some feedback for you to consider.
> > I think you should change the error message when a non-existent id is given (e.g. ‘Profile doesn’t exist’). Currently, it returns the following:
> > http://localhost:3000/api/v1/user/<non_existent_id>/get-profile
> > ```
> > {
> >     "status": "error",
> >     "message": "Cannot read property 'dataValues' of null"
> > }
> > ```
> > 
> > 
> > I find the required phone number format to be unclear -> the error message doesn’t clearly show a user the required response format. I found it a challenge to enter a Rwandan number: 256 786 123 456. Maybe you can change the regex rule here to be more flexible.
> > ```
> > {
> >     "status": "error",
> >     "message": [
> >         "phoneNumber with value \"256-786-123456\" fails to match the required pattern: /^(\\(\\d{3}\\) |\\d{3}-)\\d{3}-\\d{4}$/"
> >     ]
> > }
> > ```
> > 
> > 
> > I think you should remove the information about creating a separate profile model from the PR description since you switched to using only the User model by updating it with the required fields.
> > How will a user’s line manager be updated?
> > ```
> > {
> >     "status": "error",
> >     "message": [
> >         "\"lineManager\" is not allowed"
> >     ]
> > }
> > ```
> > 
> > 
> > I think you might have to create a separate route specifically for updating a user’s line manager since the current update profile route (PATCH /api/v1/user/update-profile) doesn’t allow you to update a user’s line manager and then maybe not allow a user to update his/her own manager.
> 
> Thanks @RaymondMwaura , let me work on these.

All these have been resolved.

timolinn commented on src/controllers/users.controller.js:
```diff
@@ -189,6 +187,57 @@ class UserController {
+      if (userData) {
+        const user = userData.dataValues;
+
+        user.lineManager = user.LineManager ? `${user.LineManager.firstName} ${user.LineManager.lastName}` : 'none';
```
seems you assigned this value here and just deleted it in the next line of code? not sure I understand why you need to do this.

gildniy replied:
No I added the object `lineManager` as the associated user, and removed the id field `lineManagerId`.

RaymondMwaura commented on src/migrations/20191112101055-create-user.js:
```diff
@@ -7,28 +7,66 @@ module.exports = {
       allowNull: false,
-      type: Sequelize.DATE
+      type: Sequelize.STRING,
+      unique: true,
```
This is a good application of database constraints to enforce the integrity of our data. 'validate:isEmail' is new for me. Good work.

RaymondMwaura reviewed (changes requested):
Everything (Postman, Terminal tests) works great. Great job. I think you still haven't addressed the issue of the 'phoneNumber' validation:

Entering: 0786123456 works
Entering: 250786123456 or +250786123456  doesn't work

For me, this is all that's remaining to be updated, I find everything else to be great.

gildniy commented:
> Everything (Postman, Terminal tests) works great. Great job. I think you still haven't addressed the issue of the 'phoneNumber' validation:
> 
> Entering: 0786123456 works
> Entering: 250786123456 or +250786123456 doesn't work
> 
> For me, this is all that's remaining to be updated, I find everything else to be great.

Resolved

b0nbon1 reviewed (changes requested):
Great work @gildniy, But I think you should  check the line manager role before assigning to the user and also if he exists.

gildniy commented:
> Great work @gildniy, But I think you should check the line manager role before assigning to the user and also if he exists.

You're right, let me add that check!

nshaphan commented:
Resolve conflicts, then this PR is ready to go!

gildniy commented:
> Resolve conflicts, then this PR is ready to go!

Let me fix these!

RaymondMwaura reviewed (changes requested):
The 'phoneNumber' validation is now great. You'll need to add the 'checkEmail' validator to ensure that invalid emails entered in the body of the request don't return a success response.

b0nbon1 reviewed (commented):
Great work @gildniy, I like you have handled the line manager validation

b0nbon1 commented on src/services/User.service.js:
```diff
@@ -11,35 +10,119 @@ class UserServices {
+    })).dataValues.id;
+
+    const isLineManagerValid = (await db.user.findOne({ where: { id: lineManagerId } })).dataValues.role === 'manager';
+
```
Sorry, I noticed this while testing that, if user updates with the profile with a line manager that does not exist it throws a `500` instead of giving out a specific message can you please handle that?? @gildniy

RaymondMwaura reviewed (approved):
Great job. All the feedback has been implemented.

b0nbon1 reviewed (approved):
Now it is working accordingly. Great job @gildniy

nshaphan reviewed (approved):
Well done @gildniy
